src/NUnitFramework/tests/Internal/Execution/ParallelExecutionTests.cs

```diff
@@ -0,0 +1,194 @@
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+#if PARALLEL
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using NUnit.Framework.Interfaces;
+using NUnit.TestData.ParallelExecutionData;
+using NUnit.TestUtilities;
+
+namespace NUnit.Framework.Internal.Execution
+{
+    [TestFixtureSource(nameof(GetParallelSuites)), Explicit]
+    public class ParallelExecutionTests : ITestListener
+    {
+        private TestSuite _testSuite;
+        private int _numCases;
+
+        private List<string> _events;
+        private TestResult _result;
+
+        private static readonly string NL = Environment.NewLine;
+
+        public ParallelExecutionTests(TestSuite testSuite, int numCases)
+        {
+            _testSuite = testSuite;
+            _numCases = numCases;
+        }
+
+        [OneTimeSetUp]
+        public void RunTestSuite()
+        {
+            _events = new List<string>();
+
+            var dispatcher = new ParallelWorkItemDispatcher(4);
+            var context = new TestExecutionContext();
+            context.Dispatcher = dispatcher;
+            context.Listener = this;
+
+            dispatcher.ShiftStarting += (shift) =>
+            {
+                AddEvent("ShiftStarted " + shift.Name);
+            };
+
+            dispatcher.ShiftFinished += (shift) =>
+            {
+                AddEvent("ShiftFinished " + shift.Name);
+            };
+
+            var workItem = TestBuilder.CreateWorkItem(_testSuite, context);
+
+            var run_complete = new ManualResetEvent(false);
+            workItem.Completed += (s, e) => { run_complete.Set(); };
+
+            dispatcher.Start(workItem);
+            run_complete.WaitOne(Timeout.Infinite);
+
+            _result = workItem.Result;
+        }
+
+        // NOTE: The following tests use Assert.Fail under control of an
+        // if statement to avoid evaluating DumpEvents unnecessarily.
+        // Unfortunately, we can't use the form of Assert that takes
+        // a Func for a message because it's not present in .NET 2.0
+
+        [Test]
+        public void AllTestsPassed()
+        {
+            if (_result.ResultState != ResultState.Success)
+                Assert.Fail(DumpEvents("Not all tests passed"));
+        }
+
+        [Test]
+        public void AllTestsRan()
+        {
+            if(_result.PassCount != _numCases)
+                Assert.Fail(DumpEvents("Incorrect number of test cases"));
+        }
+
+        [Test]
+        public void OnlyOneShiftMayBeActive()
+        {
+            int count = 0;
+            foreach (var @event in _events)
+            {
+                if (@event.StartsWith("ShiftStarted") && ++count > 1)
```

Comment:
I guess this is really not important for tests, but if this was shipping code, I'd yell about the missing `StringComparison.Ordinal` parameter because thread culture can wreck string comparisons.

Reply:
I agree with both points. 😈